src/blog/retrofitting-temporal-memory-safety-on-c++.md

```diff
@@ -0,0 +1,114 @@
+
+UAFs are often hard to spot in larger codebases where ownership of objects is transferred between various components. The general problem is so widespread that to this date both industry and academia regularly come up with mitigation strategies. The examples are endless: C++ smart pointers of all kinds are used to better define and manage ownership on application level; static analysis in compilers is used to avoid compiling problematic code in the first place; where static analysis fails, dynamic tools such as [C++ sanitizers](https://github.com/google/sanitizers) can intercept accesses and catch problems on specific executions.
+
+Chrome’s use of C++ is sadly no different here and the majority of [high-severity security bugs are UAF issues](https://www.chromium.org/Home/chromium-security/memory-safety/). In order to catch issues before they reach production, all of the aforementioned techniques are used. In addition to regular tests, fuzzers ensure that there’s always new input to work with for dynamic tools. Chrome even goes further and employs a C++ garbage collector called [Oilpan](https://v8.dev/blog/oilpan-library) which deviates from regular C++ semantics but provides temporal memory safety where used. Where such deviation is unreasonable, a new kind of smart pointer called [MiraclePtr](https://security.googleblog.com/2021/09/an-update-on-memory-safety-in-chrome.html) was introduced recently to deterministically crash on accesses to dangling pointers when used. Oilpan, MiraclePtr, and smart-pointer-based solutions require significant adoptions of the application code.
+
+Over the last decade, another approach has seen some success: memory quarantine. The basic idea is to put explicitly freed memory into quarantine and only make it available when a certain safety condition is reached. Microsoft has shipped versions of this mitigation in its browsers: [MemoryProtector](https://www.trendmicro.com/en_us/research/14/g/mitigating-uaf-exploits-with-delay-free-for-internet-explorer.html) in Internet Explorer in 2014 and its successor [MemGC](https://securityintelligence.com/memgc-use-after-free-exploit-mitigation-in-edge-and-ie-on-windows-10/) in (pre-Chromium) Edge in 2015. In the [Linux kernel](https://a13xp0p0v.github.io/2020/11/30/slab-quarantine.html) a probabilistic approach was used where memory was eventually just recycled. And this approach has seen attention in academia in recent years with the [MarkUs paper](https://www.cst.cam.ac.uk/blog/tmj32/addressing-temporal-memory-safety). The rest of this article summarizes our journey of experimenting with quarantines and heap scanning in Chrome.
+
+(At this point, one may ask where pointer authentication fits into this picture – keep on reading!)
```

Comment:
I didn't see any uses of pointer authentication, did you mean to say Memory Tagging here instead?

Reply:
I think when I wrote this I meant pointer authentication in a broad sense and not ARM's PAC mechanisms in particular.

After all, memory tagging also ensures pointer authenticity, it's just not cryptographically save. We cann call this memory tagging instead.

Reply:
Oh I see, that makes sense. It  might be a little confusing for somebody familiar with PAC, so memory tagging might be clearer indeed, but it's not that important.

Reply:
Changed to "memory tagging", thanks!